File: ports/csharp/BcbpParserTest.cs
// BcbpParserTest.cs — ชุดทดสอบ parser ฝั่ง C# (self-contained, ไม่พึ่ง framework)
//
//   cd ports/csharp
//   dotnet run                      (ถ้าตั้งเป็น console project)
//   หรือรวมกับ xUnit/NUnit ตามที่โปรเจกต์คุณใช้
//
// ตัวอย่างข้อมูลตรงกับฝั่ง Python/Java เพื่อยืนยันว่าทุกพอร์ตให้ผลเท่ากัน

using System;
using Bcbp;

class BcbpParserTest
{
    const string SingleLeg =
        "M1DESMARAIS/LUC       EABC123 YULFRAAC 0834 226F001A0025 100";

    // 2 legs + conditional (version/repeated) + security
    const string MultiLeg =
        "M2DESMARAIS/LUC       EABC123 YULFRAAC 0834 326J001A0025 162>532 RY6180 " +
        "AC                                        2A0141234567890  AC AC 1234567890" +
        "       20K DEF456 FRAGVALH 3664 227C012C0002 12C2A0140987654321  LH        " +
        "             20K ^10DABC0123456789";

    static int passed = 0, failed = 0;

    static void Check(string name, bool cond)
    {
        if (cond) { passed++; Console.WriteLine("  PASS  " + name); }
        else { failed++; Console.WriteLine("  FAIL  " + name); }
    }

    static void Eq(string name, object a, object b) =>
        Check($"{name} ({a} == {b})", Equals(a, b));

    static int Main()
    {
        var d = BcbpParser.Parse(SingleLeg);
        Eq("formatCode", d.FormatCode, "M");
        Eq("numberOfLegs", d.NumberOfLegs, 1);
        Eq("passengerName", d.PassengerName, "DESMARAIS/LUC");
        Eq("electronicTicketIndicator", d.ElectronicTicketIndicator, "E");
        var leg = d.Legs[0];
        Eq("PNR", leg.OperatingCarrierPNR, "ABC123");
        Eq("fromCity", leg.FromCity, "YUL");
        Eq("toCity", leg.ToCity, "FRA");
        Eq("operatingCarrier", leg.OperatingCarrier, "AC");
        Eq("flightNumber", leg.FlightNumber, "0834");
        Eq("seatNumber", leg.SeatNumber, "001A");
        Eq("checkInSequenceNumber", leg.CheckInSequenceNumber, "0025");

        var d24 = BcbpParser.Parse(SingleLeg, 2024);
        Eq("julianDate2024", d24.Legs[0].FlightDate, new DateTime(2024, 8, 13));

        Check("noYearHintDateNull", d.Legs[0].FlightDate == null);
        Eq("rawJulianKept", d.Legs[0].DateOfFlight, "226");

        var m = BcbpParser.Parse(MultiLeg, 2024);
        Eq("multiLegCount", m.NumberOfLegs, 2);
        Check("multiLegSize", m.Legs.Count == 2);
        Eq("leg1From", m.Legs[0].FromCity, "YUL");
        Eq("leg2From", m.Legs[1].FromCity, "FRA");
        Eq("leg2To", m.Legs[1].ToCity, "GVA");
        Check("versionPresent", m.VersionNumber != null);
        Check("securityPresent", m.Security != null);
        Eq("securityType", m.Security["type"], "1");
        Eq("ffNumberLeg1", m.Legs[0].ConditionalRepeated["frequentFlyerNumber"], "1234567890");

        foreach (var bad in new[] { "HELLO WORLD", "X1ABC" })
        {
            bool threw = false;
            try { BcbpParser.Parse(bad); } catch (BcbpParseException) { threw = true; }
            Check($"reject[{bad}]", threw);
        }

        var t = BcbpParser.Parse("M1SHORT/NAME");
        Eq("truncatedName", t.PassengerName, "SHORT/NAME");

        Console.WriteLine($"\n{passed} passed, {failed} failed");
        return failed > 0 ? 1 : 0;
    }
}

File: ports/java/BcbpParserTest.java
// BcbpParserTest.java — ชุดทดสอบ parser ฝั่ง Java (ไม่พึ่ง JUnit)
//
//   cd ports/java
//   javac BcbpParser.java BcbpParserTest.java
//   java bcbp.BcbpParserTest
//
// ตัวอย่างข้อมูลตรงกับฝั่ง Python (tests/test_parser.py) เพื่อยืนยันว่าทุกพอร์ตให้ผลเท่ากัน

package bcbp;

import java.time.LocalDate;

public class BcbpParserTest {

    static final String SINGLE_LEG =
        "M1DESMARAIS/LUC       EABC123 YULFRAAC 0834 226F001A0025 100";

    // 2 legs + conditional (version/repeated) + security
    static final String MULTI_LEG =
        "M2DESMARAIS/LUC       EABC123 YULFRAAC 0834 326J001A0025 162>532 RY6180 "
      + "AC                                        2A0141234567890  AC AC 1234567890"
      + "       20K DEF456 FRAGVALH 3664 227C012C0002 12C2A0140987654321  LH        "
      + "             20K ^10DABC0123456789";

    static int passed = 0, failed = 0;

    static void check(String name, boolean cond) {
        if (cond) { passed++; System.out.println("  PASS  " + name); }
        else { failed++; System.out.println("  FAIL  " + name); }
    }

    static void eq(String name, Object a, Object b) {
        check(name + " (" + a + " == " + b + ")", a == null ? b == null : a.equals(b));
    }

    public static void main(String[] args) {
        // --- single leg mandatory ---
        BcbpParser.Result d = BcbpParser.parse(SINGLE_LEG);
        eq("formatCode", d.formatCode, "M");
        eq("numberOfLegs", d.numberOfLegs, 1);
        eq("passengerName", d.passengerName, "DESMARAIS/LUC");
        eq("electronicTicketIndicator", d.electronicTicketIndicator, "E");
        BcbpParser.Leg leg = d.legs.get(0);
        eq("PNR", leg.operatingCarrierPNR, "ABC123");
        eq("fromCity", leg.fromCity, "YUL");
        eq("toCity", leg.toCity, "FRA");
        eq("operatingCarrier", leg.operatingCarrier, "AC");
        eq("flightNumber", leg.flightNumber, "0834");
        eq("dateOfFlight", leg.dateOfFlight, "226");
        eq("seatNumber", leg.seatNumber, "001A");
        eq("checkInSequenceNumber", leg.checkInSequenceNumber, "0025");

        // --- Julian date conversion (226 ของปี 2024 = 13 ส.ค.) ---
        BcbpParser.Result d24 = BcbpParser.parse(SINGLE_LEG, 2024);
        eq("julianDate2024", d24.legs.get(0).flightDate, LocalDate.of(2024, 8, 13));

        // --- ไม่มี yearHint => flightDate เป็น null แต่ยังเก็บ Julian ดิบ ---
        check("noYearHintDateNull", d.legs.get(0).flightDate == null);
        eq("rawJulianKept", d.legs.get(0).dateOfFlight, "226");

        // --- multi-leg ---
        BcbpParser.Result m = BcbpParser.parse(MULTI_LEG, 2024);
        eq("multiLegCount", m.numberOfLegs, 2);
        check("multiLegSize", m.legs.size() == 2);
        eq("leg1From", m.legs.get(0).fromCity, "YUL");
        eq("leg2From", m.legs.get(1).fromCity, "FRA");
        eq("leg2To", m.legs.get(1).toCity, "GVA");
        check("versionPresent", m.versionNumber != null);
        check("securityPresent", m.security != null);
        eq("securityType", m.security.get("type"), "1");
        eq("ffNumberLeg1", m.legs.get(0).conditionalRepeated.get("frequentFlyerNumber"), "1234567890");

        // --- reject non-BCBP ---
        for (String bad : new String[]{"HELLO WORLD", "X1ABC"}) {
            boolean threw = false;
            try { BcbpParser.parse(bad); } catch (BcbpParser.BcbpParseException e) { threw = true; }
            check("reject[" + bad + "]", threw);
        }

        // --- tolerate truncated ---
        BcbpParser.Result t = BcbpParser.parse("M1SHORT/NAME");
        eq("truncatedName", t.passengerName, "SHORT/NAME");

        System.out.println("\n" + passed + " passed, " + failed + " failed");
        if (failed > 0) System.exit(1);
    }
}
